### PR TITLE
feat(executors): add colocated dual-engine with vLLM sleep/wake

### PR DESCRIPTION
**Pull request by icenfly**
## Summary

Part of #127. Lays groundwork for #129.

### Background

While using RL-Kernel for a full SFT → GRPO pipeline on Qwen3 (8×A100-SXM4-80GB), I designed a 6-dimension reward function and improved the validation full-field match rate from 83.78% to 85.95% (+2.17%). During training I noticed a clear GPU utilization problem in the standard disaggregated setup (6 GPUs training + 2 GPUs vLLM): training GPUs sit idle during rollout, and inference GPUs sit idle during training — roughly 25% waste.

This PR implements a **colocated** alternative: all 8 GPUs are shared between vLLM inference and DeepSpeed training, switching via vLLM's sleep/wake mechanism (v0.8+). Sleep/wake overhead is only 3–8% of step time, far less than the 25% idle waste in disaggregated mode.

## What Changed

**Core orchestration:**
- `examples/colocated_dual_engine.py` — `ColocatedOrchestrator`: vLLM(TP=N) + persistent DeepSpeed workers on shared GPUs. Phase loop: `wake → rollout → sleep → train → offload → repeat`.
- `examples/_colocated_train_worker.py` — Persistent training worker launched once via torchrun, stays alive across all steps. Uses file-based signals for coordination and pinned CPU memory for fast GPU↔CPU weight transfer.

**Inference adapter:**
- `rl_engine/executors/inference_adapter.py` — `InferenceEngineAdapter` Protocol with `sleep()`/`wake_up()`/`generate()`/`update_weights()`. `VLLMInferenceAdapter` is fully functional; `SGLangInferenceAdapter` is a placeholder (SGLang does not yet support sleep mode).

**Tests and docs:**
- `tests/test_colocated_dual_engine.py` — Unit tests (orchestrator init, adapter creation, Protocol conformance, WeightBridge integration) + GPU integration test (vLLM sleep/wake memory lifecycle).
- `docs/guides/colocated_setup.md` — Architecture overview, Quick Start, configuration, memory budget.

### Training quality: Colocated matches Disaggregated

Both architectures were evaluated on the same validation set (968 samples, 6-dimension reward, Qwen3 SFT → GRPO).

**Step-aligned comparison (step 100):**

| Metric | Disaggregated step 100 | Colocated step 100 | Delta |
|---|---:|---:|---:|
| Full-field match rate | **85.95%** | 85.33% | -0.62% |
| JSON parse rate | 98.86% | **99.38%** | +0.52% |
| Avg output length | 275 | **215** | -22% |
| Reward mean | 2.97 | 2.92 | -0.05 |

At the same training step, colocated is 0.62% behind on match rate but produces shorter, more parseable outputs. The gap is within noise for 968 samples.

**Best checkpoint comparison:**

| Metric | Disaggregated best (step 100) | Colocated best (step 500) |
|---|---:|---:|
| Full-field match rate | **85.95%** | **85.95%** |
| JSON parse rate | 98.86% | **99.48%** |
| Avg output length | 275 | **212** |
| Reward mean | 2.97 | 2.93 |

Colocated converges slightly slower (best at step 500 vs 100) but reaches the same peak accuracy. The slower convergence is expected — colocated uses all 8 GPUs for both phases with time-sharing overhead, while disaggregated dedicates 6 GPUs exclusively to training.

### GPU efficiency

| | Disaggregated | Colocated |
|---|---|---|
| Inference GPUs | 2 dedicated (idle during training) | 8 shared |
| Training GPUs | 6 dedicated (idle during rollout) | 8 shared |
| GPU idle waste | ~25% | 0% |
| Switching overhead | N/A | 3–8% (sleep/wake) |
| Net utilization gain | baseline | **+17–22%** |

### Colocated benchmark details

| Run | Model | Steps | Avg Step | Sleep/Wake Overhead |
|---|---|---:|---:|---:|
| Persistent worker | Qwen3-0.6B | 5 | 30.0s | **7.3%** |
| Qwen3-8B | Qwen3-8B | 10 | 36.8s | **3.1%** |
| Pinned memory opt | Qwen3-0.6B | 5 | 22.9s | **8.5%** |

- Loss converges normally across all runs.
- Pinned memory optimization reduces step time by ~24%.
- No OOM with `gpu_memory_utilization=0.40`.

## Known Limitations

1. Single-node only — multi-node requires NCCL weight transport.
2. Sequential phases — no async overlap between rollout and training yet.
3. `vllm_gpu_memory_utilization` must be tuned per model size.

## Testing

```bash
# Unit tests (no GPU)
python -m pytest tests/test_colocated_dual_engine.py -v -k "not VLLMSleepWake"

# GPU integration test
python -m pytest tests/test_colocated_dual_engine.py -v -k "VLLMSleepWake"

# Smoke test
python examples/colocated_dual_engine.py \
  --model Qwen/Qwen3-0.6B --num-gpus 1 --steps 3
```

## Summary by CodeRabbit

* **New Features**
  * Added an experimental single-node setup for sharing GPUs between vLLM inference and DeepSpeed training.
  * Added inference adapters supporting vLLM lifecycle management and weight updates, with initial SGLang support.
  * Added an example workflow that alternates rollout and training phases, records metrics, and supports configurable models, GPUs, and training settings.
  * Added a persistent distributed training worker for multi-step LoRA training.

* **Documentation**
  * Added setup instructions, configuration guidance, architecture details, benchmarks, and known limitations.

* **Tests**
  * Added coverage for adapter behavior, GPU memory transitions, weight synchronization, and orchestration.

### CODE REVIEW

**Comment by icenfly** [comment]
@coderabbitai review

**Comment by Flink-ddd** [comment]
Thank you for work, but The work on ws3 is not currently in the planning stage and requires discussion and approval before a corresponding PR can be submitted.

**Comment by icenfly** [comment]
@Flink-ddd Thank you for your notice, would be happy to contribute to some planned issues and some need-gpu-cis. Looking forward for further discussion on this topic.

**Comment by Flink-ddd** [comment]
Sure, welcome to contribution for gpu-cis issue.
